src/bin/cernan.rs

```diff
@@ -50,6 +65,26 @@ fn populate_forwards(
     }
 }
 
+fn join_all(workers : HashMap<String, SinkWorker>){
+    for (_worker_id, worker) in workers {
+        worker.thread.join().expect("Failed to join worker");
+    }
+}
+
+fn broadcast_shutdown(workers : &HashMap<String, SinkWorker>){
+    let mut source_channels = Vec::new();
+    for (id, worker) in workers.iter() {
```

Comment:
@blt This broadcast approach thwarts at least once delivery. 

Consider a case where we have a topology where a sink is at the end of some chain of sinks/filters.  What we really want is to inject the SHUTDOWN at the top most sink/filter for each chain then let that SHUTDOWN event be forwarded along.

Do you happen to know a way to iterate over the root nodes of each chain?

Reply:
Right now it’s goofy because there’s no proper graph structure but we build
up the “top” of the paths for time pulses. I imagine you could reuse that
list.

Somewhat longer term we should probably invest in a proper graph. Cycle
detection would be nice.